ibis/expr/tests/test_rules.py

```diff
@@ -170,3 +170,161 @@ class MyOp(ops.ValueOp):
+                    rules.column(name='group', value_type=rules.number),
+                    rules.column(name='value', value_type=rules.number),
+                ],
+                allow_extra=allow_extra)]
+        output_type = rules.type_of_arg(0)
+
+    check_op_input(MyOp, schema, raises)
+
+
+@pytest.mark.parametrize(
+    ('schema', 'allow_extra', 'raises'),
+
+    [([('group', dt.int64),
+       ('value', dt.double)], False, False),
+
+     ([('group', dt.int64),
+       ('value', dt.double),
+       ('value2', dt.double)], False, False),
+
+     ([('group', dt.int64),
+       ('value', dt.double),
+       ('value2', dt.timestamp)], False, True),
+
+     ([('group', dt.int64),
+       ('value', dt.double),
+       ('value3', dt.timestamp)], False, True),
+
+     ([('group', dt.int64),
+       ('value', dt.double),
+       ('value2', dt.double),
+       ('value3', dt.double)], False, True),
+
+     ([('group', dt.int64),
+       ('value', dt.double),
+       ('value2', dt.double),
+       ('value3', dt.timestamp)], True, False)])
+def test_table_with_schema_optional(schema, allow_extra, raises):
+    class MyOp(ops.ValueOp):
+        input_type = [
+            rules.table(
+                name='table',
+                schema=[
+                    rules.column(name='group', value_type=rules.number),
+                    rules.column(name='value', value_type=rules.number),
+                    rules.column(name='value2', value_type=rules.number,
+                                 optional=True)],
+                allow_extra=allow_extra)]
+        output_type = rules.type_of_arg(0)
+
+    check_op_input(MyOp, schema, raises)
+
+
+@pytest.mark.parametrize(
+    ('schema', 'raises'),
+
+    [([('group', dt.int64),
+       ('value1', dt.double),
+       ('value2', dt.double),
+       ('value3', dt.double)], False)])
+def test_table_schema_and_satisfying(schema, raises):
+    class MyOp(ops.ValueOp):
+        input_type = [
+            rules.table(
+                name='table',
+                satisfying=[
+                    lambda t: len(t.columns) >= 4,
+                    lambda t: t.schema().types.count(dt.Double()) >= 3],
+                schema=[
+                    rules.column(name='group', value_type=rules.number),
+                    rules.column(name='value1', value_type=rules.number),
+                    rules.column(name='value2', value_type=rules.number),
+                    rules.column(name='value3', value_type=rules.number),
+                    rules.column(name='value4', value_type=rules.number,
+                                 optional=True)])]
+        output_type = rules.type_of_arg(0)
+
+    check_op_input(MyOp, schema, raises)
+
+
+def test_table_not_a_table():
+    class MyOp(ops.ValueOp):
+        input_type = [rules.table(name='table')]
+        output_type = rules.type_of_arg(0)
+
+    with pytest.raises(IbisTypeError):
+        MyOp(123)
+
+
+def test_table_invalid_satisfying():
```

Comment:
can you add a test that doesn't pass either satisfying or schema

Reply:
Done.